core/app/models/spree/weighted_adjustments_handler.rb

```diff
@@ -0,0 +1,47 @@
+module Spree
+  class WeightedAdjustmentsHandler
+    attr_reader :line_item, :order, :preferred_amount
+
+    def initialize(line_item, preferred_amount)
+      @line_item = line_item
+      @order = line_item.order
+      @preferred_amount = preferred_amount
+    end
+
+    # Returns the weighted adjustment for the initialized line item
+    #
+    # @return [Integer]
+    def amount
+      weighted_amounts[@line_item.id].to_f
+    end
+
+    private
+
+    # Returns a hash of line item IDs and their corresponding weighted
+    # adjustments
+    #
+    # @private
+    # @return [Hash<Integer, Integer>]
+    def weighted_amounts
+      remaining_amount = @preferred_amount
+
+      @order.line_items.each_with_index.map do |line_item, i|
```

Comment:
Not super important but could we add a `sort_by(&:id)` to this? Just to avoid the "rounding penny' from ending up on different items in different runs?